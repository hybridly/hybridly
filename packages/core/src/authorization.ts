export interface Authorizable<Authorizations extends Record<string, boolean>> {
	authorization: Authorizations
}

/**
 * Checks whether the given data has the authorization for the given action.
 * If the data object has no authorization definition corresponding to the given action, this method will return `false`.
 */
function can<
    Authorizations extends Record<string, boolean>,
    Data extends Authorizable<Authorizations>,
    Action extends keyof Data['authorization']
>(action: Action, resource: Data) {
    if (resource === undefined) {
        return useProperty('security.user').value?.authorization?.[action] ?? false
    }

    return resource.authorization?.[action] ?? false
}

function cannot<
    Authorizations extends Record<string, boolean>,
    Data extends Authorizable<Authorizations>,
    Action extends keyof Data['authorization']
>(action: Action, resource: Data) {
    return ! can(action, resource)
}
