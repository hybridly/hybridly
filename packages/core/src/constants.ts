export const STORAGE_EXTERNAL_KEY = 'hybridly:external'
export const HYBRIDLY_HEADER = 'x-hybrid'
export const EXTERNAL_NAVIGATION_HEADER = `${HYBRIDLY_HEADER}-external`
export const EXTERNAL_NAVIGATION_TARGET_HEADER = `${HYBRIDLY_HEADER}-external-target`
export const PARTIAL_COMPONENT_HEADER = `${HYBRIDLY_HEADER}-partial-component`
export const ONLY_DATA_HEADER = `${HYBRIDLY_HEADER}-only-data`
export const DIALOG_KEY_HEADER = `${HYBRIDLY_HEADER}-dialog-key`
export const DIALOG_REDIRECT_HEADER = `${HYBRIDLY_HEADER}-dialog-redirect`
export const EXCEPT_DATA_HEADER = `${HYBRIDLY_HEADER}-except-data`
export const VERSION_HEADER = `${HYBRIDLY_HEADER}-version`
export const ERROR_BAG_HEADER = `${HYBRIDLY_HEADER}-error-bag`
export const SCROLL_REGION_ATTRIBUTE = 'scroll-region'
