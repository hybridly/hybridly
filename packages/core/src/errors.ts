import type { AxiosResponse } from 'axios'

export class NotAHybridResponseError extends Error {
	constructor(public response: AxiosResponse) {
		super()
	}
}

export class NavigationCancelledError extends Error {}

export class RoutingNotInitialized extends Error {
	constructor() {
		super('Routing is not initialized. Make sure the Vite plugin is enabled and that `php artisan route:list` returns no error.')
	}
}

export class RouteNotFound extends Error {
	constructor(name: string) {
		super(`Route [${name}] does not exist.`)
	}
}

export class MissingRouteParameter extends Error {
	constructor(parameter: string, routeName: string) {
		super(`Parameter [${parameter}] is required for route [${routeName}].`)
	}
}
