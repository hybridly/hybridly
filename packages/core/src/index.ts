export { createRouter, router } from './router/router'
export type { Router, HybridRequestOptions, NavigationResponse, HybridPayload, ResolveComponent, Method, Progress } from './router/types'

export { getRouterContext } from './context'
export type { RouterContext, RouterContextOptions } from './context'

export { definePlugin, registerHook } from './plugins'
export type { Plugin } from './plugins'

export { makeUrl, sameUrls } from './url'
export type { UrlResolvable } from './url'

export { can } from './authorization'
export type { Authorizable } from './authorization'

export { route } from './routing'
export type { RouteDefinition, GlobalRouteCollection, RoutingConfiguration, RouteName, RouteParameters } from './routing'

export type { DynamicConfiguration } from './config'

export * as constants from './constants'
export * from './types'
