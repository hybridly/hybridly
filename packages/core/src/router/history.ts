import { debounce, debug } from '@hybridly/utils'
import { SCROLL_REGION_ATTRIBUTE } from '../constants'
import type { InternalRouterContext, RouterContextOptions, Serializer } from '../context'
import { getInternalRouterContext, getRouterContext, setContext } from '../context'
import { runHooks } from '../plugins'
import { saveScrollPositions } from '../scroll'
import { makeUrl } from '../url'
import { navigate } from './router'

type SerializedContext = Omit<InternalRouterContext, 'adapter' | 'serializer' | 'plugins' | 'hooks' | 'axios' | 'routes'>

/** Puts the given context into the history state. */
export function setHistoryState(options: HistoryOptions = {}) {
	if (!window?.history) {
		throw new Error('The history API is not available, so Hybridly cannot operate.')
	}

	const context = getRouterContext()
	const method = options.replace
		? 'replaceState'
		: 'pushState'

	const serialized = serializeContext(context)

	debug.history('Setting history state:', {
		method,
		context,
		serialized,
	})

	try {
		window.history[method](serialized, '', context.url)
	} catch (error) {
		console.error('Hybridly could not save its current state in the history. This is most likely due to a property being non-serializable, such as a proxy or a reference.')
		throw error
	}
}

/** Gets the current history state if it exists. */
export function getHistoryState<T = any>(key?: string): T {
	const state = key
		? window.history.state?.state?.[key]
		: window.history.state?.state

	return getRouterContext().serializer.unserialize<T>(state)
}

/** Register history-related event listeneners. */
export async function registerEventListeners() {
	const context = getRouterContext()
	debug.history('Registering [popstate] and [scroll] event listeners.')

	// Popstate is for catching back and forward navigations. We want
	// to imitate native browser behavior while keeping the SPA feeling.
	window?.addEventListener('popstate', async(event) => {
		debug.history('Navigation detected (popstate event). State:', { state: event.state })

		await runHooks('backForward', {}, event.state, context)

		// If there is no state in this history entry, we come from the user
		// replacing the URL manually. In this case, we want to restore everything
		// like it was. We need to copy the hash if any and restore the scroll positions.
		if (!event.state) {
			debug.history('There is no state. Adding hash if any and restoring scroll positions.')

			return await navigate({
				payload: {
					...context,
					url: makeUrl(context.url, { hash: window.location.hash }).toString(),
				},
				preserveScroll: true,
				preserveState: true,
				replace: true,
			})
		}

		// If the history entry has been tempered with, we want
		// to use it. We swap the components accordingly.
		await navigate({
			payload: event.state,
			preserveScroll: true,
			preserveState: !!getInternalRouterContext().dialog || !!event.state.dialog,
			updateHistoryState: false,
			isBackForward: true,
		})
	})

	// On scroll, we want to save the positions of all scrollbars.
	// This is needed in order to restore them upon navigation.
	window?.addEventListener('scroll', (event) => debounce(() => {
		if ((event?.target as Element)?.hasAttribute?.(SCROLL_REGION_ATTRIBUTE)) {
			saveScrollPositions()
		}
	}, 100), true)
}

/** Checks if the current navigation was made by going back or forward. */
export function isBackForwardNavigation(): boolean {
	if (!window.history.state) {
		return false
	}

	return (window.performance?.getEntriesByType('navigation').at(0) as PerformanceNavigationTiming)?.type === 'back_forward'
}

/** Handles a navigation which was going back or forward. */
export async function handleBackForwardNavigation(): Promise<void> {
	debug.router('Handling a back/forward navigation.')
	window.history.state.version = getRouterContext().version

	await navigate({
		payload: window.history.state,
		preserveScroll: true,
		preserveState: false,
		updateHistoryState: false,
		isBackForward: true,
	})
}

/** Saves a value into the current history state. */
export function remember<T = any>(key: string, value: T): void {
	debug.history(`Remembering key "${key}" with value`, value)

	// We avoid propagation in order to not trigger a recursive
	// render by telling the adapter we updated.
	setContext({
		state: {
			...getRouterContext().state,
			[key]: value,
		},
	}, { propagate: false })

	setHistoryState({ replace: true })
}

/** Gets a value saved into the current history state. */
export function getKeyFromHistory<T = any>(key: string): T {
	return getHistoryState<T>(key)
}

/** Serializes the context so it can be written to the history state. */
export function serializeContext(context: InternalRouterContext): SerializedContext {
	return {
		url: context.url,
		version: context.version,
		view: context.serializer.serialize(context.view),
		dialog: context.dialog,
		scrollRegions: context.scrollRegions,
		state: context.serializer.serialize(context.state),
	}
}

export function createSerializer(options: RouterContextOptions): Serializer {
	if (options.serializer) {
		return options.serializer
	}

	return {
		serialize: (view) => JSON.parse(JSON.stringify(view)),
		unserialize: (state) => state,
	}
}

export interface HistoryOptions {
	/**
	 * Replace the current history entry instead of pushing a new entry.
	 * @default false
	 */
	replace?: boolean
}
