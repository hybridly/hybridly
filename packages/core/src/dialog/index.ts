import { getInternalRouterContext } from '../context'
import type { HybridRequestOptions } from '../router'
import { performLocalNavigation, performHybridNavigation } from '../router'

export interface CloseDialogOptions extends HybridRequestOptions {
	/**
	 * Close the dialog without a round-trip to the server.
	 * @default false
	 */
	local?: boolean
}

/**
 * Closes the dialog.
 */
export async function closeDialog(options?: CloseDialogOptions) {
	const context = getInternalRouterContext()
	const url = context.dialog?.redirectUrl ?? context.dialog?.baseUrl

	if (!url) {
		return
	}

	context.adapter.onDialogClose?.(context)

	if (options?.local === true) {
		return await performLocalNavigation(url, {
			preserveScroll: true,
			preserveState: true,
			dialog: false,
			component: context.view.component,
			properties: context.view.properties,
			...options,
		})
	}

	return await performHybridNavigation({
		url,
		preserveScroll: true,
		preserveState: true,
		...options,
	})
}
