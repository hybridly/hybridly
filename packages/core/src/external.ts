import type { AxiosResponse } from 'axios'
import qs from 'qs'
import { debug } from '@hybridly/utils'
import type { HybridRequestOptions } from 'hybridly'
import { EXTERNAL_NAVIGATION_HEADER, STORAGE_EXTERNAL_KEY } from './constants'
import { getRouterContext, setContext } from './context'
import { navigate } from './router/router'
import type { UrlResolvable } from './url'
import { makeUrl, sameUrls } from './url'

/**
 * Performs an external navigation by saving options to the storage and
 * making a full page reload. Upon loading, the navigation options
 * will be pulled and a hybrid navigation will be made.
 */
export async function performExternalNavigation(options: ExternalNavigationOptions): Promise<void> {
	debug.external('Navigating to an external URL:', options)
	window.sessionStorage.setItem(STORAGE_EXTERNAL_KEY, JSON.stringify(options))
	window.location.href = options.url

	// If the external navigation is to the same page, we need to manually perform
	// a full page reload
	if (sameUrls(window.location, options.url)) {
		debug.external('Manually reloading due to the external URL being the same.')
		window.location.reload()
	}
}

/** Navigates to the given URL without the hybrid protocol. */
export function navigateToExternalUrl(url: UrlResolvable, data?: HybridRequestOptions['data']) {
	document.location.href = makeUrl(url, {
		search: qs.stringify(data, {
			encodeValuesOnly: true,
			arrayFormat: 'brackets',
		}),
	}).toString()
}

/** Checks if the response wants to redirect to an external URL. */
export function isExternalResponse(response: AxiosResponse): boolean {
	return response?.status === 409 && !!response?.headers?.[EXTERNAL_NAVIGATION_HEADER]
}

/**
 * Performs the internal navigation when an external navigation to a hybrid view has been made.
 * This method is meant to be called on router creation.
 */
export async function handleExternalNavigation(): Promise<void> {
	debug.external('Handling an external navigation.')
	const options = JSON.parse(window.sessionStorage.getItem(STORAGE_EXTERNAL_KEY) || '{}') as ExternalNavigationOptions
	window.sessionStorage.removeItem(STORAGE_EXTERNAL_KEY)

	debug.external('Options from the session storage:', options)

	// If we navigated to somewhere with a hash, we need to update the context
	// to add said hash because it was initialized without it.
	setContext({
		url: makeUrl(getRouterContext().url, { hash: window.location.hash }).toString(),
	})

	await navigate({
		type: 'initial',
		preserveState: true,
		preserveScroll: options.preserveScroll,
	})
}

/** Checks if the navigation being initialized points to an external location. */
export function isExternalNavigation(): boolean {
	try {
		return window.sessionStorage.getItem(STORAGE_EXTERNAL_KEY) !== null
	} catch {}

	return false
}

interface ExternalNavigationOptions {
	/** Target URL. */
	url: string
	/** Whether to preserve the scroll if the external navigation leads to a hybrid view. */
	preserveScroll: boolean
}
