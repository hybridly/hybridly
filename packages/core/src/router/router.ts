import type { AxiosProgressEvent, AxiosResponse } from 'axios'
import { showResponseErrorModal, match, merge, when, debug, random, hasFiles, objectToFormData } from '@hybridly/utils'
import { ERROR_BAG_HEADER, EXCEPT_DATA_HEADER, EXTERNAL_NAVIGATION_HEADER, ONLY_DATA_HEADER, PARTIAL_COMPONENT_HEADER, HYBRIDLY_HEADER, VERSION_HEADER, DIALOG_KEY_HEADER, DIALOG_REDIRECT_HEADER } from '../constants'
import { NotAHybridResponseError, NavigationCancelledError } from '../errors'
import type { InternalRouterContext, RouterContextOptions } from '../context'
import { getRouterContext, initializeContext, payloadFromContext, setContext } from '../context'
import { handleExternalNavigation, isExternalResponse, isExternalNavigation, performExternalNavigation, navigateToExternalUrl } from '../external'
import { resetScrollPositions, restoreScrollPositions, saveScrollPositions } from '../scroll'
import type { UrlResolvable } from '../url'
import { sameHashes, fillHash, makeUrl, normalizeUrl, sameUrls } from '../url'
import { runHooks } from '../plugins'
import { generateRouteFromName, getRouteDefinition } from '../routing/route'
import { closeDialog } from '../dialog'
import { setHistoryState, isBackForwardNavigation, handleBackForwardNavigation, registerEventListeners, getHistoryState, getKeyFromHistory, remember } from './history'
import type { ConditionalNavigationOption, Errors, ComponentNavigationOptions, NavigationOptions, Router, HybridRequestOptions, HybridPayload, NavigationResponse, Method } from './types'

/**
 * The hybridly router.
 * This is the core function that you can use to navigate in
 * your application. Make sure the routes you call return a
 * hybrid response, otherwise you need to call `external`.
 *
 * @example
 * router.get('/posts/edit', { post })
 */
export const router: Router = {
	abort: async() => getRouterContext().pendingNavigation?.controller.abort(),
	active: () => !!getRouterContext().pendingNavigation,
	navigate: async(options) => await performHybridNavigation(options),
	reload: async(options) => await performHybridNavigation({ preserveScroll: true, preserveState: true, ...options }),
	get: async(url, options = {}) => await performHybridNavigation({ ...options, url, method: 'GET' }),
	post: async(url, options = {}) => await performHybridNavigation({ preserveState: true, ...options, url, method: 'POST' }),
	put: async(url, options = {}) => await performHybridNavigation({ preserveState: true, ...options, url, method: 'PUT' }),
	patch: async(url, options = {}) => await performHybridNavigation({ preserveState: true, ...options, url, method: 'PATCH' }),
	delete: async(url, options = {}) => await performHybridNavigation({ preserveState: true, ...options, url, method: 'DELETE' }),
	local: async(url, options = {}) => await performLocalNavigation(url, options),
	external: (url, data = {}) => navigateToExternalUrl(url, data),
	to: async(name, parameters, options) => {
		const url = generateRouteFromName(name, parameters)
		const method = getRouteDefinition(name).method.at(0)
		return await performHybridNavigation({ url, ...options, method })
	},
	dialog: {
		close: (options) => closeDialog(options),
	},
	history: {
		get: (key) => getKeyFromHistory(key),
		remember: (key, value) => remember(key, value),
	},
}

/** Creates the hybridly router. */
export async function createRouter(options: RouterContextOptions): Promise<InternalRouterContext> {
	await initializeContext(options)

	return await initializeRouter()
}

/** Performs every action necessary to make a hybrid navigation. */
export async function performHybridNavigation(options: HybridRequestOptions): Promise<NavigationResponse> {
	const navigationId = random()
	const context = getRouterContext()
	debug.router('Making a hybrid navigation:', { context, options, navigationId })

	try {
		// Sets the method if not specifically defined.
		if (!options.method) {
			debug.router('Setting method to GET because none was provided.')
			options.method = 'GET'
		}

		// Force uppercase method because we accept lowercase methods,
		// *angry look at Hassan*
		options.method = options.method.toUpperCase() as Method

		// If applicable, converts the data to a `FormData` object.
		if ((hasFiles(options.data) || options.useFormData) && !(options.data instanceof FormData)) {
			options.data = objectToFormData(options.data)
			debug.router('Converted data to FormData.', options.data)

			// Automatically spoofs the method of the form, because it is not possible to
			// send a PUT, PATCH or DELETE request with files in them.
			if (options.method && ['PUT', 'PATCH', 'DELETE'].includes(options.method) && options.spoof !== false) {
				debug.router(`Automatically spoofing method ${options.method}.`)
				options.data.append('_method', options.method)
				options.method = 'POST'
			}
		}

		// Converts data to query parameters if the method is GET
		// and some non-FormData data is provided.
		if (!(options.data instanceof FormData) && options.method === 'GET' && Object.keys(options.data ?? {}).length) {
			debug.router('Transforming data to query parameters.', options.data)
			options.url = makeUrl(options.url ?? context.url, {
				query: options.data,
			})
			options.data = {}
		}

		// Before anything else, we fire the "before" event to make sure
		// there was no user-specified handler returning "false".
		if (!await runHooks('before', options.hooks, options, context)) {
			debug.router('"before" event returned false, aborting the navigation.')
			throw new NavigationCancelledError('The navigation was cancelled by the "before" event.')
		}

		// Abort any active navigation.
		if (context.pendingNavigation) {
			debug.router('Aborting current navigation.', context.pendingNavigation)
			context.pendingNavigation?.controller?.abort()
		}

		// Before making the navigation, we need to make sure the scroll positions are
		// saved, so we can restore them later.
		saveScrollPositions()

		// If the URL has transformation options, apply them before using the URL.
		if (options.url && options.transformUrl) {
			options.url = makeUrl(options.url, options.transformUrl)
		}

		const targetUrl = makeUrl(options.url ?? context.url)
		const abortController = new AbortController()

		// A navigation is being made, we need to add it to the context so it
		// can be reused later on.
		setContext({
			pendingNavigation: {
				id: navigationId,
				url: targetUrl,
				controller: abortController,
				status: 'pending',
				options,
			},
		})

		await runHooks('start', options.hooks, context)
		debug.router('Making request with axios.')

		const response = await context.axios.request({
			url: targetUrl.toString(),
			method: options.method,
			data: options.method === 'GET' ? {} : options.data,
			params: options.method === 'GET' ? options.data : {},
			signal: abortController.signal,
			headers: {
				...options.headers,
				...(context.dialog ? { [DIALOG_KEY_HEADER]: context.dialog!.key } : {}),
				...(context.dialog ? { [DIALOG_REDIRECT_HEADER]: context.dialog!.redirectUrl ?? '' } : {}),
				...when(options.only !== undefined || options.except !== undefined, {
					[PARTIAL_COMPONENT_HEADER]: context.view.component,
					...when(options.only, { [ONLY_DATA_HEADER]: JSON.stringify(options.only) }, {}),
					...when(options.except, { [EXCEPT_DATA_HEADER]: JSON.stringify(options.except) }, {}),
				}, {}),
				...when(options.errorBag, { [ERROR_BAG_HEADER]: options.errorBag }, {}),
				...when(context.version, { [VERSION_HEADER]: context.version }, {}),
				[HYBRIDLY_HEADER]: true,
				'X-Requested-With': 'XMLHttpRequest',
				'Accept': 'text/html, application/xhtml+xml',
			},
			validateStatus: () => true,
			onUploadProgress: async(event: AxiosProgressEvent) => {
				await runHooks('progress', options.hooks, {
					event,
					percentage: Math.round(event.loaded / (event.total ?? 0) * 100),
				}, context)
			},
		})

		await runHooks('data', options.hooks, response, context)

		// An external response is a hybrid response that wants a full page
		// load to a requested URL. It may be the same URL, in which
		// case a full page refresh will be performed.
		if (isExternalResponse(response)) {
			debug.router('The response is explicitely external.')
			await performExternalNavigation({
				url: fillHash(targetUrl, response.headers[EXTERNAL_NAVIGATION_HEADER]!),
				preserveScroll: options.preserveScroll === true,
			})

			return { response }
		}

		// An invalid response is a response that do not declare itself via
		// the protocole header.
		// In such cases, we want to throw to handler it later.
		if (!isHybridResponse(response)) {
			throw new NotAHybridResponseError(response)
		}

		// At this point, we know the response respects the hybridly protocol.
		debug.router('The response respects the Hybridly protocol.')
		const payload = response.data as HybridPayload

		// If the navigation was asking for specific properties, we ensure that the
		// new request object contains the properties of the current view context,
		// because the back-end sent back only the required properties.
		if ((options.only?.length ?? options.except?.length) && payload.view.component === context.view.component) {
			debug.router(`Merging ${options.only ? '"only"' : '"except"'} properties.`, payload.view.properties)
			payload.view.properties = merge(context.view.properties, payload.view.properties)
			debug.router('Merged properties:', payload.view.properties)
		}

		// If everything was according to the plan, we can make our navigation and
		// update the context. Underlying adapters get the updated data.
		await navigate({
			payload: {
				...payload,
				url: fillHash(targetUrl, payload.url),
			},
			preserveScroll: options.preserveScroll,
			preserveState: options.preserveState,
			preserveUrl: options.preserveUrl,
			replace: options.replace === true || sameHashes(payload.url, window.location.href) || options.preserveUrl,
		})

		// If the new view's properties has errors, userland expects an event
		// with said errors to be emitted. However, errors can be scoped with
		// an error bag, and if the given error bag is missing, the event data
		// will be empty.
		if (Object.keys(context.view.properties.errors ?? {}).length > 0) {
			const errors = (() => {
				if (options.errorBag && typeof context.view.properties.errors === 'object') {
					return (context.view.properties.errors as any)[options.errorBag] ?? {}
				}

				return context.view.properties.errors
			})() as Errors

			debug.router('The request returned validation errors.', errors)
			setContext({
				pendingNavigation: {
					...context.pendingNavigation!,
					status: 'error',
				},
			})

			await runHooks('error', options.hooks, errors, context)
		} else {
			setContext({
				pendingNavigation: {
					...context.pendingNavigation!,
					status: 'success',
				},
			})

			await runHooks('success', options.hooks, payload, context)
		}

		return { response }
	//
	} catch (error: any) {
		await match(error.constructor.name, {
			NavigationCancelledError: async() => {
				debug.router('The request was cancelled through the "before" hook.', error)
				await runHooks('abort', options.hooks, context)
			},
			AbortError: async() => {
				debug.router('The request was aborted.', error)
				await runHooks('abort', options.hooks, context)
			},
			NotAHybridResponseError: async() => {
				debug.router('The request was not hybridly.')
				console.error(error)
				await runHooks('invalid', options.hooks, error, context)
				if (context.responseErrorModals) {
					showResponseErrorModal(error.response.data)
				}
			},
			default: async() => {
				if (error?.name === 'CanceledError') {
					debug.router('The request was cancelled.', error)
					await runHooks('abort', options.hooks, context)
				} else {
					debug.router('An unknown error occured.', error)
					console.error(error)
					await runHooks('exception', options.hooks, error, context)
				}
			},
		})

		await runHooks('fail', options.hooks, context)

		return {
			error: {
				type: error.constructor.name,
				actual: error,
			},
		}
	} finally {
		debug.router('Ending navigation.')
		await runHooks('after', options.hooks, context)

		// If the navigation is pending, it means another one has started,
		// and if we clean it up, it could lead to issues after the request is done.
		if (context.pendingNavigation?.id === navigationId) {
			setContext({ pendingNavigation: undefined })
		}
	}
}

/** Checks if the response contains a hybrid header. */
export function isHybridResponse(response: AxiosResponse): boolean {
	return !!response?.headers[HYBRIDLY_HEADER]
}

/**
 * Makes an internal navigation that swaps the view and updates the context.
 * @internal This function is meant to be used internally.
 */
export async function navigate(options: NavigationOptions) {
	const context = getRouterContext()
	debug.router('Making an internal navigation:', { context, options })

	await runHooks('navigating', {}, options, context)

	// If no request was given, we use the current context instead.
	options.payload ??= payloadFromContext()

	function evaluateConditionalOption<T extends boolean | string>(option?: ConditionalNavigationOption<T>) {
		return typeof option === 'function'
			? option(options)
			: option
	}

	const shouldPreserveState = evaluateConditionalOption(options.preserveState)
	const shouldPreserveScroll = evaluateConditionalOption(options.preserveScroll)
	const shouldReplaceHistory = evaluateConditionalOption(options.replace)
	const shouldReplaceUrl = evaluateConditionalOption(options.preserveUrl)

	// If the navigation was asking to preserve the current state, we also need to
	// update the context's state from the history state.
	if (shouldPreserveState && getHistoryState() && options.payload.view.component === context.view.component) {
		debug.history('Setting the history from this entry into the context.')
		setContext({ state: getHistoryState() })
	}

	// If the navigation required the URL to be preserved, we skip its update
	// by replacing the payload URL with the current context URL.
	if (shouldReplaceUrl) {
		debug.router(`Preserving the current URL (${context.url}) instead of navigating to ${options.payload.url}`)
		options.payload!.url = context.url
	}

	// We merge the new request into the current context. That will replace
	// view, dialog, url and version, so the context is in sync with the
	// navigation that took place.
	setContext({
		...options.payload,
		state: {},
	})

	// History state must be updated to preserve the expected, native browser behavior.
	// However, in some cases, we just want to swap the views without making an
	// actual navigation. Additionally, we don't want to actually push a new state
	// when navigating to the same URL.
	if (options.updateHistoryState !== false) {
		debug.router(`Target URL is ${context.url}, current window URL is ${window.location.href}.`, { shouldReplaceHistory })
		setHistoryState({ replace: shouldReplaceHistory })
	}

	// Then, we swap the view.
	const viewComponent = await context.adapter.resolveComponent(context.view.component)
	debug.router(`Component [${context.view.component}] resolved to:`, viewComponent)
	await context.adapter.onViewSwap({
		component: viewComponent,
		dialog: context.dialog,
		properties: options.payload?.view.properties,
		preserveState: shouldPreserveState,
	})

	if (options.isBackForward) {
		restoreScrollPositions()
	} else if (!shouldPreserveScroll) {
		resetScrollPositions()
	}

	await runHooks('navigated', {}, options, context)
}

/** Initializes the router by reading the context and registering events if necessary. */
async function initializeRouter(): Promise<InternalRouterContext> {
	const context = getRouterContext()

	if (isBackForwardNavigation()) {
		handleBackForwardNavigation()
	} else if (isExternalNavigation()) {
		handleExternalNavigation()
	} else {
		debug.router('Handling the initial page navigation.')

		// If we navigated to somewhere with a hash, we need to update the context
		// to add said hash because it was initialized without it.
		setContext({
			url: makeUrl(context.url, { hash: window.location.hash }).toString(),
		})

		await navigate({
			preserveState: true,
			replace: sameUrls(context.url, window.location.href),
		})
	}

	registerEventListeners()

	await runHooks('ready', {}, context)

	return context
}

/** Performs a local navigation to the given component without a round-trip. */
export async function performLocalNavigation(targetUrl: UrlResolvable, options?: ComponentNavigationOptions) {
	const context = getRouterContext()
	const url = normalizeUrl(targetUrl)

	return await navigate({
		...options,
		payload: {
			version: context.version,
			dialog: options?.dialog === false ? undefined : (options?.dialog ?? context.dialog),
			url,
			view: {
				component: options?.component ?? context.view.component,
				properties: options?.properties ?? {},
			},
		},
	})
}
