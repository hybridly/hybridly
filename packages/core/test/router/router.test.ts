import { beforeEach, it } from 'vitest'
import { performHybridNavigation, router } from '../../src/router/router'
import { getRouterContext, registerHook } from '../../src'
import { fakeRouterContext, fakePayload, mockSuccessfulUrl } from '../utils'
import { server } from '../server'

beforeEach(async() => {
	await fakeRouterContext()
})

it('performs hybrid navigations', async({ expect }) => {
	server.resetHandlers(
		mockSuccessfulUrl({
			json: fakePayload({
				url: 'https://localhost.test/navigation',
				view: {
					component: 'target.view',
					properties: {
						foo: 'bar',
					},
				},
			}),
		}),
	)

	const { response } = await performHybridNavigation({
		url: 'http://localhost.test/navigation',
	})

	expect(response?.data).toMatchSnapshot('navigation response')
	expect(getRouterContext()).toMatchSnapshot('context after navigation')
})

it('performs external navigations', async({ expect }) => {
	router.external('http://localhost.test/navigation', {
		owo: 'uwu',
		uwu: {
			foo: 'bar',
		},
	})

	expect(document.location.href).toBe('http://localhost.test/navigation?owo=uwu&uwu[foo]=bar')
})

it('supports global "before" event cancellation', async({ expect }) => {
	const options = { url: 'http://localhost.test/navigation' }
	registerHook('before', () => false)

	expect((await performHybridNavigation(options)).error?.type).toBe('NavigationCancelledError')
})

it('supports scoped "before" event cancellation', async({ expect }) => {
	const options = {
		url: 'http://localhost.test/navigation',
		hooks: { before: () => false },
	}

	expect((await performHybridNavigation(options)).error?.type).toBe('NavigationCancelledError')
})
