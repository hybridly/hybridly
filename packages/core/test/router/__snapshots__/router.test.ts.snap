// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`performs hybrid navigations > context after navigation 1`] = `
{
  "adapter": {
    "onViewSwap": [Function],
    "resolveComponent": [Function],
    "updateRoutingConfiguration": [Function],
  },
  "axios": [Function],
  "hooks": {},
  "memo": {},
  "pendingNavigation": undefined,
  "plugins": [],
  "preloadCache": Map {},
  "responseErrorModals": undefined,
  "routing": undefined,
  "scrollRegions": [
    {
      "left": 0,
      "top": 0,
    },
    {
      "left": 0,
      "top": 0,
    },
  ],
  "serializer": {
    "serialize": [Function],
    "unserialize": [Function],
  },
  "url": "https://localhost.test/navigation",
  "version": "abc123",
  "view": {
    "component": "target.view",
    "properties": {
      "foo": "bar",
    },
  },
}
`;

exports[`performs hybrid navigations > navigation response 1`] = `
{
  "url": "https://localhost.test/navigation",
  "version": "abc123",
  "view": {
    "component": "target.view",
    "properties": {
      "foo": "bar",
    },
  },
}
`;
